File: backend/src/main/java/com/utsav/arts/services/OrdersService.java
package com.utsav.arts.services;

import com.utsav.arts.exceptions.InvalidRequestException;
import com.utsav.arts.exceptions.ResourceNotFoundException;
import com.utsav.arts.models.OrderStatus;
import com.utsav.arts.models.Orders;

import java.util.List;
import java.util.Optional;

/**
 * Service interface for managing customer orders.
 * Handles creating, updating, retrieving, and deleting orders.
 */
public interface OrdersService {

    /**
     * Places a new order for a user.
     * <p>
     * - Converts the user's cart items into order items.
     * - Calculates the total price.
     * - Sets the order status to PENDING.
     *
     * @param userId  ID of the user placing the order
     * @param address Shipping address for the order
     * @return The created order
     * @throws ResourceNotFoundException if the user or their cart does not exist
     * @throws InvalidRequestException   if the user's cart is empty
     */
    Orders placeOrder(int userId, String address);

    /**
     * Confirms an order, changing its status to CONFIRMED.
     *
     * @param orderId ID of the order to confirm
     * @return The updated order
     * @throws ResourceNotFoundException if the order does not exist
     * @throws InvalidRequestException   if the order cannot transition to CONFIRMED
     */
    Orders confirmOrder(int orderId);

    /**
     * Marks an order as shipped.
     *
     * @param orderId ID of the order to ship
     * @return The updated order
     * @throws ResourceNotFoundException if the order does not exist
     * @throws InvalidRequestException   if the order cannot transition to SHIPPED
     */
    Orders shipOrder(int orderId);

    /**
     * Marks an order as delivered.
     *
     * @param orderId ID of the order to deliver
     * @return The updated order
     * @throws ResourceNotFoundException if the order does not exist
     * @throws InvalidRequestException   if the order cannot transition to DELIVERED
     */
    Orders deliverOrder(int orderId);

    /**
     * Cancels an order.
     *
     * @param orderId ID of the order to cancel
     * @return The updated order
     * @throws ResourceNotFoundException if the order does not exist
     * @throws InvalidRequestException   if the order cannot transition to CANCELLED
     */
    Orders cancelOrder(int orderId);

    /**
     * Finds an order by its ID.
     *
     * @param id The order ID
     * @return Optional containing the order if found, otherwise empty
     */
    Optional<Orders> findById(int id);

    /**
     * Returns all orders.
     *
     * @return List of all orders
     */
    List<Orders> findAll();

    /**
     * Returns all orders for a specific user.
     *
     * @param userId The user ID
     * @return List of orders belonging to the user
     */
    List<Orders> findByUserId(int userId);

    /**
     * Returns all orders with a specific status.
     *
     * @param status The order status to filter by
     * @return List of orders with the given status
     */
    List<Orders> findByStatus(OrderStatus status);

    /**
     * Deletes an order by its ID.
     *
     * @param id ID of the order to delete
     * @throws ResourceNotFoundException if the order does not exist
     */
    void deleteById(int id);

    /**
     * Checks if a user is the owner of a given order.
     *
     * @param orderId The order ID
     * @param userId  The user ID
     * @return true if the user owns the order, false otherwise
     */
    boolean isOwner(int orderId, int userId);
}

File: frontend/src/services/orders/ordersService.ts
import api from "../api";
import type {
  OrderStatus,
  OrdersRequest,
  OrdersResponse,
} from "./orders.types";

export const ordersService = {
  /**
   * Places a new order using the current user's cart items.
   *
   * @see {@link OrdersRequest}
   * @see {@link OrdersResponse}
   */
  async placeOrder(data: OrdersRequest): Promise<OrdersResponse> {
    const response = await api.post<OrdersResponse>("/orders", data);
    return response.data;
  },

  /**
   * Retrieves a specific order by ID.
   *
   * @see {@link OrdersResponse}
   */
  async getOrderById(id: number): Promise<OrdersResponse> {
    const response = await api.get<OrdersResponse>(`/orders/${id}`);
    return response.data;
  },

  /**
   * Retrieves all orders (Admin/Owner only).
   *
   * @see {@link OrdersResponse}
   */
  async getAllOrders(): Promise<OrdersResponse[]> {
    const response = await api.get<OrdersResponse[]>("/orders");
    return response.data;
  },

  /**
   * Retrieves all orders for a specific user.
   *
   * @see {@link OrdersResponse}
   */
  async getOrdersByUserId(userId: number): Promise<OrdersResponse[]> {
    const response = await api.get<OrdersResponse[]>(`/orders/user/${userId}`);
    return response.data;
  },

  /**
   * Retrieves orders filtered by status (Admin/Owner only).
   *
   * @see {@link OrdersStatus}
   * @see {@link OrdersResponse}
   */
  async getOrdersByStatus(status: OrderStatus): Promise<OrdersResponse[]> {
    const response = await api.get<OrdersResponse[]>(
      `/orders/status/${status}`,
    );
    return response.data;
  },

  // ---------------- STATUS UPDATES ----------------

  /**
   * Confirm an order. Can be called by OWNER only.
   *
   * @see {@link OrdersResponse}
   */
  async confirmOrder(id: number): Promise<OrdersResponse> {
    const response = await api.post<OrdersResponse>(`/orders/${id}/confirm`);
    return response.data;
  },

  /**
   * Ship an order. Can be called by OWNER only.
   *
   * @see {@link OrdersResponse}
   */
  async shipOrder(id: number): Promise<OrdersResponse> {
    const response = await api.post<OrdersResponse>(`/orders/${id}/ship`);
    return response.data;
  },

  /**
   * Deliver an order. Can be called by OWNER only.
   *
   * @see {@link OrdersResponse}
   */
  async deliverOrder(id: number): Promise<OrdersResponse> {
    const response = await api.post<OrdersResponse>(`/orders/${id}/deliver`);
    return response.data;
  },

  /**
   * Cancels an order. Can be called by OWNER or the order owner.
   *
   * @see {@link OrdersResponse}
   */
  async cancelOrder(id: number): Promise<OrdersResponse> {
    const response = await api.post<OrdersResponse>(`/orders/${id}/cancel`);
    return response.data;
  },

  /**
   * Deletes an order record (Admin/Owner only).
   */
  async deleteOrder(id: number): Promise<void> {
    await api.delete(`/orders/${id}`);
  },
};
